utils/fixed_decimal/src/compact.rs

```diff
@@ -0,0 +1,154 @@
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::convert::TryFrom;
+use core::fmt;
+
+use core::str::FromStr;
+
+use crate::Error;
+use crate::FixedDecimal;
+
+/// A struct containing a [`FixedDecimal`] significand together with an exponent, representing a
+/// number written in compact notation (such as 1.2M).
+/// This represents a _source number_ that uses compact decimal notation, as defined
+/// [in UTS #35](https://www.unicode.org/reports/tr35/tr35-numbers.html#Plural_rules_syntax).
+///
+/// This is distinct from [`crate::ScientificDecimal`] because it does not represent leading 0s
+/// nor a sign in the exponent, and behaves differently in pluralization.
+#[derive(Debug, Clone, PartialEq)]
+pub struct CompactDecimal {
+    significand: FixedDecimal,
+    exponent: i16,
+}
+
+/// Render the [`CompactDecimal`] in sampleValue syntax.
+/// The letter c is used, rather than the deprecated e.
+///
+/// # Examples
+///
+/// ```
+/// # use fixed_decimal::CompactDecimal;
+/// # use std::str::FromStr;
+/// # use writeable::assert_writeable_eq;
+/// #
+/// assert_writeable_eq!(CompactDecimal::from_str("+1.20c6").unwrap(), "+1.20c6");
+/// ```
+impl writeable::Writeable for CompactDecimal {
+    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        self.significand.write_to(sink)?;
+        sink.write_char('c')?;
+        self.exponent.write_to(sink)
+    }
+
+    fn writeable_length_hint(&self) -> writeable::LengthHint {
+        self.significand.writeable_length_hint() + 1 + self.exponent.writeable_length_hint()
+    }
+}
+
+writeable::impl_display_with_writeable!(CompactDecimal);
+
+impl FromStr for CompactDecimal {
+    type Err = Error;
+    fn from_str(input_str: &str) -> Result<Self, Self::Err> {
+        Self::try_from(input_str.as_bytes())
+    }
+}
+
+/// The deprecated letter e is not accepted as a synonym for c.
+impl TryFrom<&[u8]> for CompactDecimal {
+    type Error = Error;
+    fn try_from(input_str: &[u8]) -> Result<Self, Self::Error> {
+        if input_str.iter().any(|&c| c == b'e' || c == b'E') {
+            return Err(Error::Syntax);
+        }
+        let mut parts = input_str.split(|&c| c == b'c');
+        let significand = FixedDecimal::try_from(parts.next().ok_or(Error::Syntax)?)?;
+        let exponent_str = parts.next().ok_or(Error::Syntax)?;
+        if parts.next().is_some() {
+            return Err(Error::Syntax);
+        }
+        if exponent_str.is_empty()
+            || exponent_str.first() == Some(&b'0')
+            || !exponent_str.iter().all(|c| (b'0'..=b'9').contains(c))
+        {
+            return Err(Error::Syntax);
+        }
+        #[allow(clippy::unwrap_used)] // All ASCII digits, so valid UTF-8.
+        let exponent = core::str::from_utf8(exponent_str)
+            .unwrap()
+            .parse()
+            .map_err(|_| Error::Limit)?;
```

Comment:
Suggestion: I think you can get nearly identical behavior with just

```suggestion
        if exponent_str.first() == Some(&b'0') {
            return Err(Error::Syntax);
        }
        let exponent = core::str::from_utf8(exponent_str)
            .ok_or(Error::Syntax)?
            .parse()
            .map_err(|_| Error::Limit)?;
```

Reply:
An approach like that conflates the overflow with the syntax error (so an overly large exponent would be reported a syntax error), which isn’t ideal. More importantly I am very uncomfortable with the extent to which it depends on exactly what `parse` accepts.

For instance, you also want to reject signs. It turns out that `parse` doesn’t accept `_` grouping marks nor leading spaces, but this is [not really documented](https://doc.rust-lang.org/std/primitive.i8.html#impl-FromStr-for-i8) (it is documented for [`from_str_radix`](https://doc.rust-lang.org/std/primitive.i8.html#method.from_str_radix) though).

Since we are implementing a specific syntax, I think it makes sense to explicitly check it.

Reply:
Reordered the checks, as discussed.